<a href="https://colab.research.google.com/github/marhoayan/Ds220Web/blob/main/Math_484_Final_Project_Simplex_Algorithm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Setting Up The Jupyter Notebook**

In [ ]:
from sympy import init_session

In [ ]:
init_session()

IPython console for SymPy 1.13.1 (Python 3.10.12-64-bit) (ground types: python)

These commands were executed:
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at https://docs.sympy.org/1.13.1/



Creating The Big M

In [ ]:
M = symbols('M',integer =True ,positive =True)

# **Feeding Python The Initial Tableau**

In [ ]:
T = Matrix([[-20*M,-5-15*M,5+15*M,2+3*M,2,-2,4-10*M,M,M,-M,-M,M,-M,0,0,0,0,0,0],
          [2,2,-2,-2,-1,1,3,-1,0,0,0,0,0,1,0,0,0,0,0],
          [1,2,-2,-4,7,-7,6,0,-1,0,0,0,0,0,1,0,0,0,0],
          [2,5,-5,4,-8,8,1,0,0,1,0,0,0,0,0,1,0,0,0],
          [4,3,-3,-6,1,-1,2,0,0,0,1,0,0,0,0,0,1,0,0],
          [8,1,-1,2,-3,3,7,0,0,0,0,-1,0,0,0,0,0,1,0],
          [3,2,-2,3,4,-4,-9,0,0,0,0,0,1,0,0,0,0,0,1]])


In [ ]:
T

⎡-20⋅M  -15⋅M - 5  15⋅M + 5  3⋅M + 2  2   -2  4 - 10⋅M  M   M   -M  -M  M   -M  0  0  0  0  0  0⎤
⎢                                                                                               ⎥
⎢  2        2         -2       -2     -1  1      3      -1  0   0   0   0   0   1  0  0  0  0  0⎥
⎢                                                                                               ⎥
⎢  1        2         -2       -4     7   -7     6      0   -1  0   0   0   0   0  1  0  0  0  0⎥
⎢                                                                                               ⎥
⎢  2        5         -5        4     -8  8      1      0   0   1   0   0   0   0  0  1  0  0  0⎥
⎢                                                                                               ⎥
⎢  4        3         -3       -6     1   -1     2      0   0   0   1   0   0   0  0  0  1  0  0⎥
⎢                                                                                               ⎥
⎢  8        1       

Recall: My choice for the initial basic variables was ${y_1,y_2,y_3,y_4,y_5,y_6}$

# **First Pivot**

The reduced cost in column 1 is negative: $(x^+_1 = -15M-5)$. I choose column 1 as my pivot column. The variable $x^+_1$ will enter the basis.

Now let's compute the ratios $\frac{x_b(i)}{u(i)}$ for each
$u_i$

In [ ]:
[ T[i,0] / T[i,1] for i in range (1,7) ]

[1, 1/2, 2/5, 4/3, 8, 3/2]

In [ ]:
[0. + T[i,0] / T[i,1] for i in range (1,7) ]

[1.0, 0.5, 0.4, 1.33333333333333, 8.0, 1.5]

The minimum ratio is $\frac{2}{5}$. It occurs in row 3. I choose row 3 as my pivot row. $x_b(3) = y_3$ will exit the basis.


Now, we need to compute the elementary row operations of the pivot.


Replace Row 0 by itself plus $\frac{15M+5}{5}$ times the pivot row(i.e row 3)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 0,k=(15*M+5)/5,row2 = 3)

Replace Row 1 by itself plus $\frac{-2}{5}$ times the pivot row(i.e row 3)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 1,k=-Rational(2,5),row2 = 3)

Replace Row 2 by itself plus $\frac{-2}{5}$ times the pivot row(i.e row 3)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 2,k=-Rational(2,5),row2 = 3)

Replace Row 4 by itself plus $\frac{-3}{5}$ times the pivot row(i.e row 3)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 4,k=-Rational(3,5),row2 = 3)

Replace Row 5 by itself plus $\frac{-1}{5}$ times the pivot row(i.e row 3)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 5,k=-Rational(1,5),row2 = 3)

Replace Row 6 by itself plus $\frac{-2}{5}$ times the pivot row(i.e row 3)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 6,k=-Rational(2,5),row2 = 3)

Multiply the pivot row(row 3) by $\frac{1}{5}$

In [ ]:
T = T . elementary_row_op(op = 'n->kn',row = 3,k = Rational(1,5))

Updated tableau:

In [ ]:
T

⎡2 - 14⋅M  0  0   15⋅M + 6  -24⋅M - 6  24⋅M + 6  5 - 7⋅M  M   M   2⋅M + 1  -M  M   -M  0  0  3⋅M + ↪
⎢                                                                                                  ↪
⎢  6/5     0  0    -18/5      11/5      -11/5     13/5    -1  0    -2/5    0   0   0   1  0   -2/5 ↪
⎢                                                                                                  ↪
⎢  1/5     0  0    -28/5      51/5      -51/5     28/5    0   -1   -2/5    0   0   0   0  1   -2/5 ↪
⎢                                                                                                  ↪
⎢  2/5     1  -1    4/5       -8/5       8/5       1/5    0   0     1/5    0   0   0   0  0    1/5 ↪
⎢                                                                                                  ↪
⎢  14/5    0  0    -42/5      29/5      -29/5      7/5    0   0    -3/5    1   0   0   0  0   -3/5 ↪
⎢                                                                                          

New basic variables: ${y_1,y_2,x^+_1,y_4,y_5,y_6}$

# **Second Pivot**

The reduced cost in column 4 is negative: $x^+_3$ $= -24M-6$. I choose column 4 as my pivot column.

The variable $x^+_3$ will enter the basis.

Now let's compute the ratios $\frac{x_b(i)}{u(i)}$ for each
$u_i$.

I compute the ratios $\frac{x_b(i)}{u(i)}$ for $i =1,2,4$ $and$ $6$ but not for $i = 3$ $and$ $5$ because $u_1,u_2,u_4$ $and$ $u_6$ are positive while $u_3$ $and$ $u_5$ are negative.

In [ ]:
[T[i,0] / T[i,4] for i in [1,2,4,6]]

⎡            14  11⎤
⎢6/11, 1/51, ──, ──⎥
⎣            29  36⎦

In [ ]:
[0. + T[i,0] / T[i,4] for i in [1,2,4,6]]

[0.545454545454545, 0.0196078431372549, 0.482758620689655, 0.305555555555556]

The minimum ratio is $\frac{1}{51}$ and it occurs in row 2.

I choose row 2 as my pivot row. $x_b(2)$ $= y_2$ will exit the basis.

Now, we need to compute the elementary row operations of the pivot.

Replace Row 0 by itself plus $\frac{(40M+10)}{17}$ times the pivot row(i.e row 2)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 0,k=Rational(40,17)*M+Rational(10,17),row2 = 2)

Replace Row 1 by itself plus $(\frac{-11}{51})$ times the pivot row(i.e row 2)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 1,k=-Rational(11,51),row2 = 2)

Replace Row 3 by itself plus $(\frac{8}{51})$ times the pivot row(i.e row 2)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 3,k=Rational(8,51),row2 = 2)

Replace Row 4 by itself plus $(\frac{-29}{51})$ times the pivot row(i.e row 2)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 4,k=-Rational(29,51),row2 = 2)

Replace Row 5 by itself plus $(\frac{7}{51})$ times the pivot row(i.e row 2)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 5,k=Rational(7,51),row2 = 2)

Replace Row 6 by itself plus $(\frac{-36}{51})$ times the pivot row(i.e row 2)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 6,k=-Rational(36,51),row2 = 2)

Multiply the pivot row(row 2) by $(\frac{5}{51})$

In [ ]:
T = T . elementary_row_op(op = 'n->kn',row = 2,k = Rational(5,51))

Updated Tableau:

In [ ]:
T

⎡36   230⋅M         31⋅M   46         105⋅M   141        23⋅M   10  18⋅M   13                 40⋅M ↪
⎢── - ─────  0  0   ──── + ──  0  0   ───── + ───  M   - ──── - ──  ──── + ──  -M  M   -M  0  ──── ↪
⎢17    17            17    17          17     17          17    17   17    17                  17  ↪
⎢                                                                                                  ↪
⎢    59               -122                71               11         -16                       -1 ↪
⎢    ──      0  0     ─────    0  0       ──       -1      ──         ────     0   0   0   1    ── ↪
⎢    51                51                 51               51          51                        5 ↪
⎢                                                                                                  ↪
⎢                     -28                 28                                                       ↪
⎢   1/51     0  0     ────     1  -1      ──       0      -5/51       -2/51    0   0   0   

New basic variables: ${y_1,x^+_3,x^+_1,y_4,y_5,y_6}$

# Third Pivot

The reduced cost in column 8 is negative: $S_2$ = $\frac{-23M}{17}$-$\frac{10}{17}$. I choose column 8 as my pivot column.

The variable $S_2$ will enter the basis.

Now let's compute the ratios $\frac{x_b(i)}{u(i)}$ for each
$u_i$


I compute the ratio $\frac{x_b(i)}{u(i)}$ for $i =1,4,6$ because $u_1,u_4,u_6$ are positive while $u_2, u_3$ $and$ $u_5$ are negative.

In [ ]:
[T[i,0] / T[i,8] for i in [1,4,6]]

⎡59  137  35⎤
⎢──, ───, ──⎥
⎣11  29   12⎦

In [ ]:
[0. + T[i,0] / T[i,8] for i in [1,4,6]]

[5.36363636363636, 4.72413793103448, 2.91666666666667]

The minimum ratio is $\frac{35}{12}$ and it occurs in row 6.
I choose row 6 as my pivot row. $x_b(6)$ $= y_6$ will exit the basis.

Now, we need to compute the elementary row operations of the pivot.

Replace Row 0 by itself plus $\frac{23M}{12}+\frac{5}{6}$ times the pivot row(i.e row 6)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 0,k=Rational(23,12)*M+Rational(5,6),row2 = 6)

Replace Row 1 by itself plus $\frac{-11}{36}$ times the pivot row(i.e row 6)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 1,k=-Rational(11,36),row2 = 6)

Replace Row 2 by itself plus $\frac{5}{36}$ times the pivot row(i.e row 6)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 2,k=Rational(5,36),row2 = 6)

Replace Row 3 by itself plus $\frac{2}{9}$ times the pivot row(i.e row 6)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 3,k=Rational(2,9),row2 = 6)

Replace Row 4 by itself plus $\frac{-29}{36}$ times the pivot row(i.e row 6)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 4,k=-Rational(29,36),row2 = 6)

Replace Row 5 by itself plus $\frac{7}{36}$ times the pivot row(i.e row 6)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 5,k=Rational(7,36),row2 = 6)

Multiply the pivot row(row 6) by $(\frac{17}{12})$

In [ ]:
T = T . elementary_row_op(op = 'n->kn',row = 6,k = Rational(17,12))

Updated Tableau:

In [ ]:
T

⎡23   115⋅M         145⋅M   43           233⋅M   17         5⋅M   2          11⋅M   5         11⋅M ↪
⎢── - ─────  0  0   ───── + ──  0  0   - ───── - ──  M   0  ─── + ─  -M  M   ──── + ─  0  M   ──── ↪
⎢6     12            12     6             12     6           6    3           12    6          6   ↪
⎢                                                                                                  ↪
⎢    19               -145                 197                                 -11                 ↪
⎢    ──      0  0     ─────     0  0       ───       -1  0   -5/18   0   0     ────    1  0    -5/ ↪
⎢    36                36                  36                                   36                 ↪
⎢                                                                                                  ↪
⎢    11                                    -47                                                     ↪
⎢    ──      0  0      7/36     1  -1      ────      0   0   -1/18   0   0     5/36    0  0

New basic variables: ${y_1,x^+_3,x^+_1,y_4,y_5,S_2}$

# Fourth Pivot

The reduced cost in column 6 is negative: $x_4$ = $\frac{-233M}{12}$-$\frac{17}{6}$. I choose column 6 as my pivot column.

The variable $x_4$ will enter the basis.

Now let's compute the ratios $\frac{x_b(i)}{u(i)}$ for each
$u_i$


I compute the ratio $\frac{x_b(i)}{u(i)}$ for $i =1,4,5$ because $u_1,u_4,u_5$ are positive while $u_2, u_3$ $and$ $u_6$ are negative.

In [ ]:
[T[i,0] / T[i,6] for i in [1,4,5]]

⎡19   37   289⎤
⎢───, ───, ───⎥
⎣197  323  179⎦

The minimum ratio is $\frac{19}{197}$ and it occurs in row 1.
I choose row 1 as my pivot row. $x_b(1)$ $= y_1$ will exit the basis.

Now, we need to compute the elementary row operations of the pivot.

Replace Row 0 by itself plus $\frac{699M}{197}+\frac{102}{197}$ times the pivot row(i.e row 1)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 0,k=Rational(699,197)*M+Rational(102,197),row2 = 1)

Replace Row 2 by itself plus $\frac{47}{197}$ times the pivot row(i.e row 1)





In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row =2,k=Rational(47,197),row2 = 1)

Replace Row 3 by itself plus $\frac{68}{197}$ times the pivot row(i.e row 1)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row =3,k=Rational(68,197),row2 = 1)

Replace Row 4 by itself plus $\frac{-323}{197}$ times the pivot row(i.e row 1)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row =4,k=-Rational(323,197),row2 = 1)

Replace Row 5 by itself plus $\frac{-179}{197}$ times the pivot row(i.e row 1)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row =5,k=-Rational(179,197),row2 = 1)

Replace Row 6 by itself plus $\frac{681}{197}$ times the pivot row(i.e row 1)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row =6,k=Rational(681,197),row2 = 1)

Multiply the pivot row(row 1) by $(\frac{36}{197})$

In [ ]:
T = T . elementary_row_op(op = 'n->kn',row = 1,k = Rational(36,197))

Updated Tableau:

In [ ]:
T

⎡809   1519⋅M         1001   435⋅M              502⋅M   102     103   30⋅M          133   33⋅M  69 ↪
⎢─── - ──────  0  0   ──── - ─────  0  0   0  - ───── - ───  0  ─── - ────  -M  M   ─── - ────  ── ↪
⎢197    197           197     197                197    197     197   197           197   197    1 ↪
⎢                                                                                                  ↪
⎢    19                  -145                     -36              -10                 -11         ↪
⎢    ───       0  0      ─────      0  0   1      ────       0     ────     0   0      ────        ↪
⎢    197                  197                     197              197                 197         ↪
⎢                                                                                                  ↪
⎢    85                  -151                     -47              -24                 13          ↪
⎢    ───       0  0      ─────      1  -1  0      ────       0     ────     0   0      ─── 

New basic variables: ${x_4,x^+_3,x^+_1,y_4,y_5,S_2}$

# Fifth Pivot

The reduced cost in column 3 is negative: $x_4$ = $\frac{1001}{197}$-$\frac{435M}{197}$. I choose column 3 as my pivot column.

The variable $x_2$ will enter the basis.

Now let's compute the ratios $\frac{x_b(i)}{u(i)}$ for each
$u_i$


I only compute the ratio $\frac{x_b(i)}{u(i)}$ for $i =5$ because $u_5$is positive while $u_1,u_2,u_3,u_4$ $and$ $u_6$ are negative.

In [ ]:
[T[i,0] / T[i,3] for i in [5]]

⎡1487⎤
⎢────⎥
⎣1011⎦

The minimum ratio is $\frac{1487}{1011}$ and it occurs in row 5.
Therefore, I choose row 5 as my pivot row. $x_b(5)$ $= y_5$ will exit the basis.

Now, we need to compute the elementary row operations of the pivot.

Replace Row 0 by itself plus $\frac{435M}{1011}-\frac{1001}{1011}$ times the pivot row(i.e row 5)


In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 0,k=Rational(435,1011)*M-Rational(1001,1011),row2 = 5)

Replace Row 1 by itself plus $\frac{145}{1011}$ times the pivot row(i.e row 5)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 1,k=Rational(145,1011),row2 = 5)

Replace Row 2 by itself plus $\frac{151}{1011}$ times the pivot row(i.e row 5)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 2,k=Rational(151,1011),row2 = 5)

Replace Row 3 by itself plus $\frac{55}{1011}$ times the pivot row(i.e row 5)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 3,k=Rational(55,1011),row2 = 5)

Replace Row 4 by itself plus $\frac{576}{1011}$ times the pivot row(i.e row 5)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 4,k=Rational(576,1011),row2 = 5)

Replace Row 6 by itself plus $\frac{1249}{1011}$ times the pivot row(i.e row 5)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 6,k=Rational(1249,1011),row2 = 5)

Multiply the pivot row(row 5) by $(\frac{197}{1011})$

In [ ]:
T = T . elementary_row_op(op = 'n->kn',row = 5,k = Rational(197,1011))

Updated Tableau:

In [ ]:
T

⎡  1504⋅M   3404                        727⋅M   1433     554    55⋅M      192⋅M   1001  12⋅M   70  ↪
⎢- ────── - ────  0  0   0  0  0   0  - ───── - ────  0  ──── - ────  -M  ───── + ────  ──── + ─── ↪
⎢   337     1011                         337    1011     1011   337        337    1011  337    337 ↪
⎢                                                                                                  ↪
⎢     1192                                 -53              -55              -145                  ↪
⎢     ────        0  0   0  0  0   1       ────       0     ────      0      ─────        4/337    ↪
⎢     1011                                 1011             1011             1011                  ↪
⎢                                                                                                  ↪
⎢     1576                                -104              -127             -151          46      ↪
⎢     ────        0  0   0  1  -1  0      ─────       0     ─────     0      ─────         

New basic variables: ${x_4,x^+_3,x^+_1,y_4,x_2,S_2}$

# **Sixth Pivot**

The reduced cost in column 7 is negative: $x_4$ = $\frac{-727M}{337}$-$\frac{1433}{1011}$. I choose column 7 as my pivot column.

The variable $S_1$ will enter the basis.

Now let's compute the ratios $\frac{x_b(i)}{u(i)}$ for each
$u_i$


I compute the ratio $\frac{x_b(i)}{u(i)}$ for $i =4,5$ because $u_4,u_5$are positive while $u_1,u_2,u_3$ $and$ $u_6$ are negative.

In [ ]:
[T[i,0] / T[i,7] for i in [4,5]]

⎡1504  1487⎤
⎢────, ────⎥
⎣727   179 ⎦

In [ ]:
[0. + T[i,0] / T[i,7] for i in [4,5]]

[2.0687757909216, 8.3072625698324]

The minimum ratio is $\frac{1504}{727}$ and it occurs in row 4.
Therefore, I choose row 4 as my pivot row. $x_b(4)$ $= y_4$ will exit the basis.

Now, we need to compute the elementary row operations of the pivot.

Replace Row 0 by itself plus $\frac{2181M}{2181}+\frac{1433}{2181}$ times the pivot row(i.e row 4)


In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 0,k=Rational(2181,2181)*M+Rational(1433,2181),row2 = 4)

Replace Row 1 by itself plus $\frac{53}{2181}$ times the pivot row(i.e row 4)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 1,k=Rational(53,2181),row2 = 4)

Replace Row 2 by itself plus $\frac{104}{2181}$ times the pivot row(i.e row 4)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 2,k=Rational(104,2181),row2 = 4)

Replace Row 3 by itself plus $\frac{299}{2181}$ times the pivot row(i.e row 4)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 3,k=Rational(299,2181),row2 = 4)

Replace Row 5 by itself plus $\frac{-179}{2181}$ times the pivot row(i.e row 4)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 5,k=-Rational(179,2181),row2 = 4)

Replace Row 6 by itself plus $\frac{2360}{2181}$ times the pivot row(i.e row 4)

In [ ]:
T = T . elementary_row_op(op = 'n->n+km',row = 6,k=Rational(2360,2181),row2 = 4)

Multiply the pivot row(row 4) by $(\frac{337}{727})$

In [ ]:
T = T . elementary_row_op(op = 'n->kn',row = 4,k = Rational(337,727))

Updated Tableau:

In [ ]:
T

⎡-316                              1429   1433    1343    134               1429      1433      13 ↪
⎢─────  0  0   0  0  0   0  0  0   ────   ────    ────    ───   M   M   M + ────  M + ────  M - ── ↪
⎢ 727                              2181   2181    2181    727               2181      2181      21 ↪
⎢                                                                                                  ↪
⎢ 936                             -110     53    -343                    -110        53       343  ↪
⎢ ───   0  0   0  0  0   1  0  0  ─────   ────   ─────   8/727  0   0    ─────      ────      ──── ↪
⎢ 727                             2181    2181   2181                    2181       2181      2181 ↪
⎢                                                                                                  ↪
⎢1288                             -257    104    -385     98             -257       104       385  ↪
⎢────   0  0   0  1  -1  0  0  0  ─────   ────   ─────    ───   0   0    ─────      ────   

New basic variables: ${x_4,x^+_3,x^+_1,S_1,x_2,S_2}$


# **The Algorithm Ends**

All reduced costs are now nonnegative. The current basis is optimal and the algorithm stops. All artificial variables are at level zero.

The top left entry of the tableau tells us that:

Optimal objective: $\frac{316}{727}$

The other numbers in the zeroth column of the tableau are the basic components of the optimal basic feasible solution found by the Big-M method: $x_b(1)$$=x_4=\frac{936}{727}$, $x_b(2)=x^+_3=\frac{1288}{727}$, $x_b(3)=x^+_1=\frac{1522}{727}$, $x_b(4)=S_1=\frac{1504}{727}$, $x_b(5)=x_2=\frac{803}{727}$, $x_b(6)=S_2=\frac{13737}{727}$

Optimal basic feasible solution: ($\frac{1522}{727},0,$$\frac{803}{727},$$\frac{1288}{727},0,$$\frac{936}{727}$, $\frac{1504}{727}$, $\frac{13737}{727}$$,0,0,0,0,0,0,0,0,0,0)$